### PR TITLE
ActivatorUtilities.CreateInstance: allocation optimization

### LINKED ISSUE #693: Improve performance of ActivatorUtilities.CreateInstance
Currently, it's 30x slower than invoking factory delegate. 
While it could never reach the performance of invoking the factory delegate there is a lot of possible improvements (especially around allocations) in `ActivatorUtilities.CreateInstance`

### PR DESCRIPTION
**Pull request by Yves57**
Some optimizations to try to fix #693.

Before:
```
|          Method |        Mean |     Error |    StdDev |         Op/s |  Gen 0 | Allocated |
|---------------- |------------:|----------:|----------:|-------------:|-------:|----------:|
|  CreateInstance | 1,317.58 ns | 5.6777 ns | 4.7411 ns |    758,965.3 | 0.0267 |     784 B |
```

After:
```
|          Method |        Mean |      Error |     StdDev |         Op/s |  Gen 0 | Allocated |
|---------------- |------------:|-----------:|-----------:|-------------:|-------:|----------:|
|  CreateInstance | 1,275.97 ns | 26.4718 ns | 44.9511 ns |    783,719.0 | 0.0134 |     448 B |
```

### CODE REVIEW

File: src/Shared/src/ActivatorUtilities/ActivatorUtilities.cs [hunk 1]
```diff
@@ -327,18 +329,16 @@ private static bool TryCreateParameterMap(ParameterInfo[] constructorParameters,
             return true;
         }
 
-        private class ConstructorMatcher
+        private struct ConstructorMatcher
         {
             private readonly ConstructorInfo _constructor;
             private readonly ParameterInfo[] _parameters;
             private readonly object[] _parameterValues;
-            private readonly bool[] _parameterValuesSet;
```
Comment (Yves57) [inline src/Shared/src/ActivatorUtilities/ActivatorUtilities.cs]: `GetTypeInfo().IsAssignableFrom(null)` returns `false` (around line 357), so I don't see why we need `parameterValuesSet`.

**Comment by anurse** [comment]
Thoughts @pakrym ? @davidfowl ?

**Comment by pakrym** [comment]
Looks reasonable.

**Comment by anurse** [comment]
Thanks for the contribution @Yves57 !

**Pull request merged**